File: Tests/src/PublicAPITests/LODTests.cpp
/*
 * Copyright 2023 Magnopus LLC

 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "AssetSystemTestHelpers.h"
#include "CSP/Common/Array.h"
#include "CSP/Systems/Assets/AssetSystem.h"
#include "CSP/Systems/SystemsManager.h"
#include "SpaceSystemTestHelpers.h"
#include "TestHelpers.h"
#include "UserSystemTestHelpers.h"

#include "gtest/gtest.h"
#include <Awaitable.h>
#include <filesystem>

namespace
{

bool RequestPredicate(const csp::systems::ResultBase& result) { return result.GetResultCode() != csp::systems::EResultCode::InProgress; }

void GetLODChain(csp::systems::AssetSystem* assetSystem, const csp::systems::AssetCollection& assetCollection, csp::systems::LODChain& outLodChain)
{
    auto [Result] = AWAIT_PRE(assetSystem, GetLODChain, RequestPredicate, assetCollection);

    EXPECT_EQ(Result.GetResultCode(), csp::systems::EResultCode::Success);

    outLodChain = Result.GetLODChain();

    EXPECT_EQ(outLodChain.AssetCollectionId, assetCollection.Id);
}

void RegisterAssetToLODChain(
    csp::systems::AssetSystem* assetSystem, const csp::systems::AssetCollection& assetCollection, csp::systems::Asset& asset, int lodLevel)
{
    auto [Result] = AWAIT_PRE(assetSystem, RegisterAssetToLODChain, RequestPredicate, assetCollection, asset, lodLevel);

    EXPECT_EQ(Result.GetResultCode(), csp::systems::EResultCode::Success);

    asset = Result.GetAsset();
}

} // namespace

CSP_PUBLIC_TEST(CSPEngine, LODTests, GetEmptyLODChainTest)
{
    SetRandSeed();

    auto& systemsManager = csp::systems::SystemsManager::Get();
    auto* userSystem = systemsManager.GetUserSystem();
    auto* spaceSystem = systemsManager.GetSpaceSystem();
    auto* assetSystem = systemsManager.GetAssetSystem();

    const char* testSpaceName = "CSP-UNITTEST-SPACE-MAG";
    const char* testSpaceDescription = "CSP-UNITTEST-SPACEDESC-MAG";
    const char* testAssetCollectionName = "CSP-UNITTEST-ASSETCOLLECTION-MAG";

    char uniqueSpaceName[256];
    SPRINTF(uniqueSpaceName, "%s-%s", testSpaceName, GetUniqueString().c_str());

    char uniqueAssetCollectionName[256];
    SPRINTF(uniqueAssetCollectionName, "%s-%s", testAssetCollectionName, GetUniqueString().c_str());

    // Log in
    csp::common::String userId;
    LogInAsNewTestUser(userSystem, userId);

    // Create space
    csp::systems::Space space;
    CreateSpace(
        spaceSystem, uniqueSpaceName, testSpaceDescription, csp::systems::SpaceAttributes::Private, nullptr, nullptr, nullptr, nullptr, space);

    // Create collection
    csp::systems::AssetCollection assetCollection;
    CreateAssetCollection(assetSystem, space.Id, nullptr, uniqueAssetCollectionName, nullptr, nullptr, assetCollection);

    // Get LOD chain
    csp::systems::LODChain chain;
    GetLODChain(assetSystem, assetCollection, chain);

    EXPECT_TRUE(chain.LODAssets.IsEmpty());

    // Cleanup
    DeleteAssetCollection(assetSystem, assetCollection);
    DeleteSpace(spaceSystem, space.Id);
    LogOut(userSystem);
}

CSP_PUBLIC_TEST(CSPEngine, LODTests, RegisterAssetsToLODChainTest)
{
    SetRandSeed();

    auto& systemsManager = csp::systems::SystemsManager::Get();
    auto* userSystem = systemsManager.GetUserSystem();
    auto* spaceSystem = systemsManager.GetSpaceSystem();
    auto* assetSystem = systemsManager.GetAssetSystem();

    const char* testSpaceName = "CSP-UNITTEST-SPACE-MAG";
    const char* testSpaceDescription = "CSP-UNITTEST-SPACEDESC-MAG";
    const char* testAssetCollectionName = "CSP-UNITTEST-ASSETCOLLECTION-MAG";
    const char* testAssetName1 = "CSP-UNITTEST-ASSET1-MAG";
    const char* testAssetName2 = "CSP-UNITTEST-ASSET2-MAG";

    char uniqueSpaceName[256];
    SPRINTF(uniqueSpaceName, "%s-%s", testSpaceName, GetUniqueString().c_str());

    char uniqueAssetCollectionName[256];
    SPRINTF(uniqueAssetCollectionName, "%s-%s", testAssetCollectionName, GetUniqueString().c_str());

    char uniqueAssetName1[256];
    SPRINTF(uniqueAssetName1, "%s-%s", testAssetName1, GetUniqueString().c_str());

    char uniqueAssetName2[256];
    SPRINTF(uniqueAssetName2, "%s-%s", testAssetName2, GetUniqueString().c_str());

    // Log in
    csp::common::String userId;
    LogInAsNewTestUser(userSystem, userId);

    // Create space
    csp::systems::Space space;
    CreateSpace(
        spaceSystem, uniqueSpaceName, testSpaceDescription, csp::systems::SpaceAttributes::Private, nullptr, nullptr, nullptr, nullptr, space);

    // Create collection
    csp::systems::AssetCollection assetCollection;
    CreateAssetCollection(assetSystem, space.Id, nullptr, uniqueAssetCollectionName, nullptr, nullptr, assetCollection);

    // Create assets
    csp::systems::Asset asset1;
    CreateAsset(assetSystem, assetCollection, uniqueAssetName1, nullptr, nullptr, asset1);

    csp::systems::Asset asset2;
    CreateAsset(assetSystem, assetCollection, uniqueAssetName2, nullptr, nullptr, asset2);

    // Register to LOD chain
    RegisterAssetToLODChain(assetSystem, assetCollection, asset1, 0);
    RegisterAssetToLODChain(assetSystem, assetCollection, asset2, 1);

    // Get LOD chain
    csp::systems::LODChain chain;
    GetLODChain(assetSystem, assetCollection, chain);

    EXPECT_EQ(chain.LODAssets.Size(), 2);

    EXPECT_EQ(chain.LODAssets[0].Level, 0);
    EXPECT_EQ(chain.LODAssets[0].Asset.Id, asset1.Id);

    EXPECT_EQ(chain.LODAssets[1].Level, 1);
    EXPECT_EQ(chain.LODAssets[1].Asset.Id, asset2.Id);

    // Cleanup
    DeleteAsset(assetSystem, assetCollection, asset1);
    DeleteAsset(assetSystem, assetCollection, asset2);

    DeleteAssetCollection(assetSystem, assetCollection);
    DeleteSpace(spaceSystem, space.Id);
    LogOut(userSystem);
}

File: Tests/CSharp/src/Tests/LODTests.cs
using System;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices;

using Common = Csp.Common;
using Systems = Csp.Systems;

using CSharpTests;
using static CSharpTests.TestHelper;

namespace CSPEngine
{
    static class LODTests
    {
        static void GetLODChain(Systems.AssetSystem assetSystem, Systems.AssetCollection assetCollection, out Systems.LODChain LODChain)
        {
            using var result = assetSystem.GetLODChain(assetCollection).Result;
            var resCode = result.GetResultCode();

            Assert.AreEqual(resCode, Systems.EResultCode.Success);

            LODChain = result.GetLODChain();
        }

        static void RegisterAssetToLODChain(Systems.AssetSystem assetSystem, Systems.AssetCollection assetCollection, Systems.Asset asset, int LODLevel)
        {
            using var result = assetSystem.RegisterAssetToLODChain(assetCollection, asset, LODLevel).Result;
            var resCode = result.GetResultCode();

            Assert.AreEqual(resCode, Systems.EResultCode.Success);
        }

#if RUN_ALL_UNIT_TESTS || RUN_LOD_TESTS || RUN_LOD_GET_EMPTY_LODCHAIN_TEST
         [Test]
        public static void GetEmptyLODChainTest()
        {
            GetFoundationSystems(out var userSystem, out var spaceSystem, out var assetSystem, out _, out _, out _, out _, out _, out _, out _);

            string testSpaceName = GenerateUniqueString("OLY-UNITTEST-SPACE-REWIND");
            string testSpaceDescription = "OLY-UNITTEST-SPACEDESC-REWIND";
            string testAssetCollectionName = GenerateUniqueString("OLY-UNITTEST-ASSETCOLLECTION-REWIND");

            // Log in
            _ = UserSystemTests.LogIn(userSystem);

            // Create space
            var space = SpaceSystemTests.CreateSpace(spaceSystem, testSpaceName, testSpaceDescription, Systems.SpaceAttributes.Private, null, null, null);

            // Create asset collection
            Systems.AssetCollection assetCollection;
            AssetSystemTests.CreateAssetCollection(assetSystem, space, null, testAssetCollectionName, null, null, out assetCollection);

            // Get LOD chain
            Systems.LODChain chain;
            GetLODChain(assetSystem, assetCollection, out chain);

            Assert.AreEqual(chain.LODAssets.Size(), 0u);
        }
#endif

#if RUN_ALL_UNIT_TESTS || RUN_LOD_TESTS || RUN_LOD_REGISTER_ASSETS_TO_LODCHAIN_TEST
        public static void RegisterAssetsToLODChainTest()
        {
            GetFoundationSystems(out var userSystem, out var spaceSystem, out var assetSystem, out _, out _, out _, out _, out _, out _, out _);

            string testSpaceName = GenerateUniqueString("OLY-UNITTEST-SPACE-REWIND");
            string testSpaceDescription = "OLY-UNITTEST-SPACEDESC-REWIND";
            string testAssetCollectionName = GenerateUniqueString("OLY-UNITTEST-ASSETCOLLECTION-REWIND");
            string testAssetName1 = GenerateUniqueString("OLY-UNITTEST-ASSET1-REWIND");
            string testAssetName2 = GenerateUniqueString("OLY-UNITTEST-ASSET2-REWIND");

            // Log in
            _ = UserSystemTests.LogIn(userSystem);

            // Create space
            var space = SpaceSystemTests.CreateSpace(spaceSystem, testSpaceName, testSpaceDescription, Systems.SpaceAttributes.Private, null, null, null);

            // Create asset collection
            Systems.AssetCollection assetCollection;
            AssetSystemTests.CreateAssetCollection(assetSystem, space, null, testAssetCollectionName, null, null, out assetCollection);

            // Create assets
            Systems.Asset asset1;
            AssetSystemTests.CreateAsset(assetSystem, assetCollection, testAssetName1, null, null, out asset1);

            Systems.Asset asset2;
            AssetSystemTests.CreateAsset(assetSystem, assetCollection, testAssetName2, null, null, out asset2);

            // Register to LOD chain
            RegisterAssetToLODChain(assetSystem, assetCollection, asset1, 0);
            RegisterAssetToLODChain(assetSystem, assetCollection, asset2, 1);

            // Get LOD chain
            Systems.LODChain chain;
            GetLODChain(assetSystem, assetCollection, out chain);

            Assert.AreEqual(chain.LODAssets.Size(), 2u);

            Assert.AreEqual(chain.LODAssets[0].Level, 0);
            Assert.AreEqual(chain.LODAssets[0].Asset.Id, asset1.Id);

            Assert.AreEqual(chain.LODAssets[1].Level, 1);
            Assert.AreEqual(chain.LODAssets[1].Asset.Id, asset2.Id);
        }
#endif
    }

}